### PR TITLE
[flang][OpenMP] Fix common block missing symbol crash

### LINKED ISSUE #65034: [flang][OpenMP] Crash observed in threadprivatised common block
If a common block symbol is not used in a threadprivatised directive, a LLVM symbol error is thrown.

```
  program sample
  integer:: x
  integer :: y
  common /blk/ x
  !$omp threadprivate (/blk/)
  !$omp parallel sections copyin(/blk/)
    !$omp section
      y = y + 1
  !$omp end parallel sections
 end program sample
```

Reproducer command: `flang-new -fc1 -fopenmp -emit-fir <file_name>.f90 -o -`

```
LLVM ERROR: symbol not mapped
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: flang-new -fc1 -fopenmp -emit-fir sample.f90 -o -
#0 0x0000555557cccf20 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/nimish/community/llvm-project/install/bin/flang-new+0x2778f20)
#1 0x0000555557cca584 SignalHandler(int) Signals.cpp:0:0
#2 0x00007ffff7a37520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
#3 0x00007ffff7a8ba7c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
#4 0x00007ffff7a8ba7c __pthread_kill_internal ./nptl/pthread_kill.c:78:10
#5 0x00007ffff7a8ba7c pthread_kill ./nptl/pthread_kill.c:89:10
#6 0x00007ffff7a37476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
#7 0x00007ffff7a1d7f3 abort ./stdlib/abort.c:81:7
#8 0x000055555645d004 llvm::RISCVISAInfo::updateImplication() (.cold) RISCVISAInfo.cpp:0:0
#9 0x0000555557c2371e (/home/nimish/community/llvm-project/install/bin/flang-new+0x26cf71e)
#10 0x0000555558041e6b (/home/nimish/community/llvm-project/install/bin/flang-new+0x2aede6b)
#11 0x0000555558095041 (anonymous namespace)::FirConverter::copyHostAssociateVar(Fortran::semantics::Symbol const&, mlir::OpBuilder::InsertPoint*) Bridge.cpp:0:0
#12 0x0000555558399523 std::_Function_handler<void (Fortran::parser::OmpClause::Copyin const*, Fortran::parser::CharBlock const&), ClauseProcessor::processCopyin() const::'lambda0'(Fortran::parser::OmpClause::Copyin const*, Fortran::parser::CharBlock const&)>::_M_invoke(std::_Any_data const&, Fortran::parser::OmpClause::Copyin const*&&, Fortran::parser::CharBlock const&) OpenMP.cpp:0:0
#13 0x000055555839fa29 ClauseProcessor::processCopyin() const (/home/nimish/community/llvm-project/install/bin/flang-new+0x2e4ba29)
#14 0x00005555583ae4fb genParallelOp(Fortran::lower::AbstractConverter&, Fortran::lower::pft::Evaluation&, mlir::Location, Fortran::parser::OmpClauseList const&, bool) OpenMP.cpp:0:0
```

### PR DESCRIPTION
**Pull request by NimishMishra**
Fixes #65034 by skipping copy of host-association information if the concerned symbol is missing from the inner construct

### CODE REVIEW

**Comment by NimishMishra** [comment]
The `copyHostAssociateVar` function captures two symbol boxes: one from `lookupOneLevelUpSymbol` which fetches the original copy of the variable, and one from `shallowLookupSymbol` which fetches the copy that intends to mask the original variable.

However, in case of common block, because of the abstraction involved, there is a chance that `shallowLookupSymbol` fails to find a symbol box because relevant symbol mapping has not been created. For instance,

```
  program sample
  integer:: x
  integer :: y
  common /blk/ x
  !$omp threadprivate (/blk/)
  !$omp parallel sections copyin(/blk/)
    !$omp section
      y = y + 1
  !$omp end parallel sections
 end program sample
```

Dumping the parse-tree for parallel sections reveals that `x` is not explicitly used in the construct, thereby it being missing from the symbol map of the construct. Now when copyin executes, it fails to find `x` in the inner construct's symbol map, thereby erring out a `LLVM ERROR: symbol not mapped`. This PR fixes the same. It only executes the relevant copyin functionality iff it is sure that the symbol exists in the symbol map of the inner construct (i.e. the relevant variable is used in the inner construct).

**Comment by NimishMishra** [comment]
Ping for review!

File: flang/lib/Lower/Bridge.cpp [hunk 1]
```diff
@@ -588,6 +588,13 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         std::nullopt);
   }
 
+  bool isPresentShallowLookup(Fortran::semantics::Symbol &sym) override final {
+    Fortran::lower::SymbolBox sb = shallowLookupSymbol(sym);
+    if (shallowLookupSymbol(sym))
```
Comment (shraiysh) [inline flang/lib/Lower/Bridge.cpp:593]: Why is this called twice? And shouldn't this function just be 
```
bool isPresentShallowLookup(Fortran::semantics::Symbol &sym) {
  return shallowLookupSymbol(sym)
}
```
And in that case do we need to make this function, or can we directly call this shallowLookup() function at the use site?

Comment (NimishMishra) [reply flang/lib/Lower/Bridge.cpp:593]: `shallowLookupSymbol()` does not seem to be exposed to any interface we have access to while lowering. So I added this interface in AbstractConverter. I'll recheck once though.

And will modify `isPresentShallowLookup` is at all it needs to be there.

Comment (NimishMishra) [reply flang/lib/Lower/Bridge.cpp:589]: @shraiysh I checked this. `shallowLookupSymbol` is operating upon a symbol map `localSymbols` defined and maintained in `Bridge.cpp`. So in any case, we need to allow an interface to access the same during lowering OpenMP clauses.

**Review by shraiysh — APPROVED** [review]
LG.

**Pull request merged**